### ISSUE TITLE
How to update the terrain while the game is running?

### ISSUE THREAD

**Issue by nan0m** [Zylann/godot_heightmap_plugin#391]
Is it possible to have the changes to the terrain be reflected in the running game as well as the editor?

**Comment by Zylann** [comment]
No, also I don't know what kind of changes you are referring to. Changes made in game will not apply to the editor, and changes in the editor will not apply to the game while it is running.

**Comment by nan0m** [comment]
I was hoping that changes made in the editor could apply to the game. 
My use case is that I there is jumping in my game and if I currently have a chasm that is too broad or two ledges that are too far apart. I have to exit the game change it and restart the game. This does take a bit long and it is a thing I have to do constantly. So I was hoping to somehow have editor changes be reflected in the game at runtime

**Comment by Zylann** [comment]
You could perhaps try reloading the terrain in game after saving the map, by listening to a hotkey and calling the loading functions of HTerrainData or something like that. Haven't tested this but I imagine you'd do:

```gdscript
terrain.data.load_data(directory_path)
terrain.update_collider()
```

I don't know how to actually reliably do that automatically, as most Godot's in-game editing functionality is a core feature that I'm not sure can be extended easily. Alternatively the data folder/resource could be unassigned and reassigned which might cause the terrain to reload, though I'm not sure the collider will update.

**Comment by nan0m** [comment]
Thanks a lot for your detailed and helpful answer!
I will reconsider using placeholders. For now, reloading the terrain works with the method you proposed. However, it seems to be reloaded from the cache so the changes from the editor aren't actually brought over to the game just as you said.

I have actually figured out how to reload a terrain from disk for my own Terrain Implementation (which I am not using anymore, because it's lacking an editor and is just not as advanced). 


```gdscript
var height_image: Image
var height_texture: Texture2D

func _process(delta):
	if file_has_changed(file_path):
		await get_tree().create_timer(0.3).timeout #for safety
		load_height_image_from_path(file_path)
		refresh_height_texture()

func load_height_image_from_path(path: String) -> Image:
	var img = Image.new()
	img.load(path)
	img.convert(Image.FORMAT_RF)
	img.resize(world_size, world_size)
	return img

func refresh_height_texture():
	height_texture= ImageTexture.create_from_image(height_img)
	mat.set("shader_parameter/HeightMap", heightmap)
```
I am checking whether the time the file was last modified has changed to check whether the terrain should be updated. By separating out the `Image` from the `ImageTexture` I have been able to get the changed image from disk instead of the cache. 

I couldn't figure out how to bring that approach over to the hterrain plugin, as how images are stored and loaded is quite advanced and complex at least to me.

**Comment by Zylann** [comment]
> I couldn't figure out how to bring that approach over to the hterrain plugin, as how images are stored and loaded is quite advanced and complex at least to me.

Once you have an image, you could apply the same logic as for generating a terrain: https://hterrain-plugin.readthedocs.io/en/latest/#modifying-terrain-from-script

Eventually I wonder if it would be possible to use the `CacheMode.CACHE_IGNORE/REPLACE` enum in `ResourceLoader.load`: https://docs.godotengine.org/en/stable/classes/class_resourceloader.html#class-resourceloader-method-load
Both for HTerrainData and the resources it loads.

**Comment by Zylann** [comment]
I added a `reload` function to `HTerrainData` in 6786bca3018d49f9e30ad8abfdacf3eca4ebe82b

So now you can edit the terrain, save, tab in/out to make sure Godot imports texture modifications, and then do something like:

```gdscript
terrain.data.reload()
terrain.update_collider()
```

**Comment by nan0m** [comment]
I tested it and it works very well for me! Thank you so much!
